text/0000-typed-array-conveniences.md

```diff
@@ -0,0 +1,312 @@
+    const TYPE_TAG: /* hidden */;
+}
+```
+
+#### `neon::types::binary::Region`
+
+The `Region` struct defines a region of an underlying buffer. For performance, it's not actually checked to be a valid region until actually constructing a typed array object. (Otherwise, every method on the type would have to re-validate every time it's called, because of the possibility of the buffer getting detached.)
+
+```rust
+#[derive(Clone, Copy)]
+struct Region<'cx, T: Binary> { /* hidden */ }
+
+impl<'cx, T: Binary> Region<'cx, T> {
+    // Constructs a new typed array for this region.
+    pub fn to_typed_array<'c, C>(self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
+    where C: Context<'c>;
+}
+```
+
+## Convenient queries
+
+This RFC also defines a number of methods that query typed arrays for metadata.
+
+### `JsTypedArray`
+
+```rust
+impl<T: Binary> JsTypedArray<T> {
+    // Returns the region of the backing buffer for this typed array.
+    pub fn to_region<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> Region<'cx, T>
+    where
+        C: Context<'cx>;
+
+    // Returns the backing buffer.
+    pub fn buffer<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> Handle<'cx, JsArrayBuffer>
+    where
+        C: Context<'cx>;
+
+    // Returns the byte offset into the backing buffer.
+    pub fn byte_offset<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> usize
+    where
+        C: Context<'cx>;
+
+    // Returns the number of elements in the typed array.
+    pub fn len<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> usize
+    where
+        C: Context<'cx>;
+}
+```
+
+### `TypedArray`
+
+This RFC adds a convenience method to the `TypedArray` trait:
+
+```rust
+pub trait TypedArray: Sealed {
+    /* as before... */
+
+    // Returns the number of bytes in this typed array.
+    fn byte_length<'cx, C>(&self, cx: &mut C) -> usize
+    where
+        C: Context<'cx>;
+}
+```
+
+### `Region`
+
+The `Region` struct also defines a number of convenient queries.
+
+```rust
+impl<'cx, T: Binary> Region<'cx, T> {
+
+pub fn buffer(self) -> Handle<'cx, JsArrayBuffer>;
+
+pub fn byte_offset(self) -> usize;
+
+pub fn len(self) -> usize;
+
+pub fn byte_length(self) -> usize;
+}
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There aren't any real drawbacks to exposing this functionality, since it's provided by Node-API. There are, however, a number of tradeoffs in the particulars of the design.
+
+The main drawbacks of this RFC's proposed design are:
+
+## Naming inconsistency: `len()` vs `byte_length()`
+
+This RFC goes with `len()` to match a strong Rust convention, and `byte_length()` to match the typed array naming convention (using snake case, of course). These does lead to an inconsistency between the two. A couple alternatives would be `byte_len()` or `size()`. But then `byte_len()` ends up feeling a bit inconsistent with `byte_offset()`. We could then choose `byte_off()` but that's a bit opaque.
```

Comment:
I somewhat prefer `byte_len` and `byte_offset` to `byte_length`. I agree that `byte_off` is confusing.

Alternatively, `byte_length` could be omitted completely since it's pretty easy to implement.

```
let byte_length = std::mem::size_of::<u32>() * arr.len(&mut cx);
```

The problem with this code is it's really easy for the generic in `size_of` to fall out of sync with the type of `T` in `arr`.

Reply:
I could go with `byte_len` and `byte_offset`. But then we just had a discussion and really liked the cleanness of:

```rust
offset()
len()
size()
```